mcstatus/motd.py

```diff
@@ -0,0 +1,549 @@
+
+    Bases on this field, you can easily write your own MOTD-to-something parser.
+    """
+    bedrock: bool = False
+    """Is server Bedrock Edition? Some details may change in work of this class."""
+
+    @classmethod
+    def parse(cls, raw: Union[dict, list, str], *, bedrock: bool = False) -> Self:
+        """Parse a raw MOTD to less raw MOTD (``Motd.parsed`` field).
+
+        :param raw: Raw MOTD, directly from server.
+        :param bedrock: Is server Bedrock Edition? Nothing changes here, just sets attribute.
+        :returns: New ``MOTD`` instance.
+        """
+        if isinstance(raw, str):
+            return cls(cls._parse_as_str(raw, bedrock=bedrock), bedrock)
+
+        if isinstance(raw, list):
+            raw = {"extra": raw}
+
+        return cls(cls._parse_as_dict(raw), bedrock)
+
+    @classmethod
+    def _parse_as_dict(cls, item: dict, *, auto_add: Optional[list[ParsedMotdItem]] = None) -> list[ParsedMotdItem]:
+        """Parse a MOTD when it's dict.
+
+        :param item: ``dict`` directly from the server.
+        :param auto_add: Values to add on this item.
+            Most time, this is ``Formatting`` from top level.
+        :returns: ``ParsedMotdItem`` list, which need to be passed to ``__init__``.
+        """
+        parsed_motd: list[ParsedMotdItem] = auto_add if auto_add is not None else []
+
+        if item.get("color"):
+            try:
+                parsed_motd.append(MinecraftColor[item["color"].upper()])
+            except KeyError:
+                if item["color"] == "reset":
+                    # minecraft server actually can't return {"reset": True}, instead
+                    # it will return {"color": "reset"}. but for follow naming, we
+                    # will not move `Formatting.RESET` to `MinecraftColor`.
+                    #
+                    # see https://wiki.vg/Chat#Shared_between_all_components, `color` field
+                    parsed_motd.append(Formatting.RESET)
+                else:
+                    parsed_motd.append(WebColor.parse(hex=item["color"]))
+
+        for style_key, style_val in Formatting.__members__.items():
+            if item.get(style_key.lower()) is False or item.get(style_key.lower()) == "false":
+                parsed_motd.remove(style_val)
+            elif item.get(style_key.lower()) is not None:
+                parsed_motd.append(style_val)
+
+        text = item.get("text")
+        if text is not None:
+            parsed_motd.append(text)
+        translate = item.get("translate")
+        if text is None and translate is not None:
+            parsed_motd.append(TranslateString(translate))
+        parsed_motd.append(Formatting.RESET)
+
+        if item.get("extra"):
+            auto_add = list(filter(lambda e: type(e) is Formatting and e != Formatting.RESET, parsed_motd))
+
+            for element in item["extra"]:
+                parsed_motd.extend(cls._parse_as_dict(element, auto_add=auto_add.copy()))
```

Comment:
Instead of an additional `auto_add` argument just for recursion, why not extend the `parsed_motd` with this directly in the for loop?

```suggestion
            prefix = list(filter(lambda e: type(e) is Formatting and e != Formatting.RESET, parsed_motd))

            for element in item["extra"]:
                parsed_motd.extend(prefix)
                parsed_motd.extend(cls._parse_as_dict(element))
```

Reply:
Because we need to modify already existing `parsed_motd`. [See this test](https://github.com/py-mine/mcstatus/pull/335/files#diff-679e4b136f587ef265574883a72c32daccb5ddc0cf3b4250e2219fe1f294bd9aR145-R157).

Reply:
Ah I see, makes sense then.